src/storage/compression/compression.cpp

```diff
@@ -455,48 +454,80 @@ void IntegerBitpacking<T>::setValuesFromUncompressed(const uint8_t* srcBuffer, o
-        auto value = ((T*)srcBuffer)[posInSrc + i];
-        // Null values will usually be 0, which will not be able to be stored if there is a non-zero
-        // offset However we don't care about the value stored for null values Currently they will
-        // be mangled by storage+recovery (underflow in the subtraction below)
-        KU_ASSERT((nullMask && nullMask->isNull(posInSrc + i)) ||
-                  canUpdateInPlace(std::span(&value, 1), metadata));
-        chunk[posInChunk] = (U)(value - (T)header.offset);
-        if (posInChunk + 1 >= CHUNK_SIZE && i + 1 < numValues) {
-            fastpack(chunk, chunkStart, header.bitWidth);
-            chunkStart = (uint8_t*)getChunkStart(dstBuffer, posInDst + i + 1, header.bitWidth);
-            fastunpack(chunkStart, chunk, header.bitWidth);
-            posInChunk = 0;
-        } else {
-            posInChunk++;
-        }
-    }
-    fastpack(chunk, chunkStart, header.bitWidth);
+    const size_t lastAlignedChunkStartIdx =
+        std::max(posInDst, (posInDst + numValues) - (posInDst + numValues) % CHUNK_SIZE);
+    if (lastAlignedChunkStartIdx > posInDst) {
+        auto posInChunk = posInDst % CHUNK_SIZE;
+        U chunk[CHUNK_SIZE];
+        // TODO(bmwinger): Now I'm a bit confused here. Why do we always need to unpack the chunk
+        // first?
+        //      Can't we just unpack the chunk if we need to modify only partial of it?
+        fastunpack(chunkStart, chunk, header.bitWidth);
+        for (offset_t i = 0; i < numValues && posInDst + i < lastAlignedChunkStartIdx; i++) {
+            auto value = ((T*)srcBuffer)[posInSrc + i];
+            // Null values will usually be 0, which will not be able to be stored if there is a
+            // non-zero offset However we don't care about the value stored for null values
+            // Currently they will be mangled by storage+recovery (underflow in the subtraction
+            // below)
+            KU_ASSERT((nullMask && nullMask->isNull(posInSrc + i)) ||
+                      canUpdateInPlace(std::span(&value, 1), metadata));
+            chunk[posInChunk] = (U)(value - (T)header.offset);
+            if (posInChunk + 1 >= CHUNK_SIZE && i + 1 < numValues &&
+                posInDst + i + 1 < lastAlignedChunkStartIdx) {
+                fastpack(chunk, chunkStart, header.bitWidth);
+                chunkStart = (uint8_t*)getChunkStart(dstBuffer, posInDst + i + 1, header.bitWidth);
+                fastunpack(chunkStart, chunk, header.bitWidth);
+                posInChunk = 0;
+            } else {
+                posInChunk++;
+            }
+        }
+        fastpack(chunk, chunkStart, header.bitWidth);
+    }
+
+    U tmpChunk[CHUNK_SIZE];
+    const size_t unalignedValuesToPack = (posInDst + numValues) - lastAlignedChunkStartIdx;
+    copyValuesToTempChunkWithOffset(reinterpret_cast<const U*>(srcBuffer) + posInSrc +
+                                        lastAlignedChunkStartIdx - posInDst,
```

Comment:
expression is a bit long, probably add a local variable somewhere to have the code better express what's going on